data-access-layer/iaas/AliClient.js

```diff
@@ -0,0 +1,203 @@
+        });
+    });
+  }
+
+
+  list(container, options) {
+    if (arguments.length < 2) {
+      options = container;
+      container = this.containerName;
+    }
+    return this.storage
+      .useBucket(container)
+      .list(options)
+      .then(listOfFiles => {
+        let list = listOfFiles.objects;
+        let isTruncated = listOfFiles.isTruncated;
+        let marker = listOfFiles.nextMarker;
+        const files = [];
+        _.each(list, file => files.push(_
+          .chain(file)
+          .pick('name', 'lastModified')
+          .set('isTruncated', isTruncated)
+          .set('marker', marker)
+          .value()
+        ));
+        return files;
+      });
+  }
+
+  listFilenames(prefix, max_keys) {
+    const options = {
+      prefix: prefix,
+      'max-keys': max_keys
+    };
+
+    let fileList = [];
+    let level = 0;
+
+    function fetchFiles() {
+      level++;
+      logger.debug(`Fetching recursively at level : ${level}`);
+      const promise = new Promise(function (resolve, reject) {
+        Promise.try(() => list(this.containerName, options))
+          .then(files => {
+            logger.debug('list of files recieved - ', files);
+            if (files && files.length > 0) {
+              fileList = fileList.concat(files);
+              if (files[0].isTruncated === true && level < 10) {
+                options.marker = files[files.length - 1].marker;
+                return fetchFiles()
+                  .then(() => resolve())
+                  .catch(err => reject(err));
+              }
+            }
+            logger.debug('end of recursion');
+            resolve();
+          })
+          .catch(err => reject(err));
+      });
+      return promise;
+    }
+
+    return fetchFiles()
+      .then(() =>
+        _
+          .chain(fileList)
+          .map(file => file)
+          .value());
+  }
+
+  remove(container, file) {
+    if (arguments.length < 2) {
+      file = container;
+      container = this.containerName;
+    }
+
+    logger.info('Deleting file ' + file + ' from container ' + container);
+    return this.storage
+      .useBucket(container)
+      .delete(file)
+      .then(() => {
+        logger.info('Deleted file ' + file + ' from container ' + container);
+      })
+      .catch(err => {
+        logger.error(err.message);
+      });
+  }
+
+  download(options) {
+    return Promise.try(() => {
+      return this.storage
+        .useBucket(options.container)
+        .get(options.remote)
+        .then(result => {
+          return result.content;
+        })
+        .catch(err => {
+          console.error(err);
+        });
+    });
+  }
+
+  upload(options, buffer) {
+    return Promise.try(() => {
+      return this.storage
+        .useBucket(options.container)
+        .put(options.remote, buffer)
+        .then(() => {
+          JSON.parse(buffer);
+        })
+        .catch(err => {
+          logger.error(err);
```

Comment:
Again, keep catch behaviour same, I think it is throwing the exception in other iaas [GcpClient.js](https://github.com/cloudfoundry-incubator/service-fabrik-broker/blob/fb03166586a47686caea9101e851cacaa3219b70/data-access-layer/iaas/GcpClient.js#L179)

Reply:
Improved the catch behaviour.